src/map.rs

```diff
@@ -61,6 +62,12 @@ matcher_map!(
         "bc",
         matchers::app::is_llvm
     ),
+    (
+        MatcherType::APP,
+        "application/javascript",
+        "js",
+        matchers::no_match
+    ),
```

Comment:
I am using `Infer` in a web context where files have both a name (with extension) and streamed content (from network or local disk). When sniffing mime the priority is given to the content, and I fallback to file extension if the match fails. For some types, such Javascript or CSS, the sniffing can only be done using the file's extension.

I was thinking that infer would address both cases. If i have to use `mime_guess` in addition this is ok, but it seems to be unmaintained. I have [a PR](https://github.com/abonander/mime_guess/pull/60) to fix a critical panic opened there for a month by the way.

Also if we decided that file extension matching is out of Infer scope then we should probably remove the `Infer::get_from_ext` method that I introduced in #16. On the other side If we decide that file extension matching is within `Infer` scope then I have several enhancements planned. For instance we could handle mime type aliases as well as file extension aliases.

What do you think?

Reply:
I think functionality like `mime_guess`'s could fit into the infer crate, or at least we could argue it would have a similar api. Maybe the builtin matchers would be in a giant `match` instead of an Iterator but still...

I also use both `infer` and `mime_guess`, and I could benefit by having something like custom matchers for extensions, since my infer function currently looks like something like this:

```rust
pub fn infer<P: AsRef<Path>>(data: &[u8], path: P) -> String {
    // using infer from the master branch
    if let Some(kind) = infer::get(&data) {
        // look at the comment at the bottom on why I have to use String
        return kind.mime_type().into();
    }

    let path = path.as_ref();
    // this is a bit ugly
    let ext = path
        .extension()
        .and_then(|ext| ext.to_str())
        .unwrap_or_default()
        .to_lowercase();
    match ext.as_str() {
        // some RAW Image formats
        "arw" => "image/x-sony-arw".into(),
        "cr2" => "image/x-canon-cr2".into(),
        "crw" => "image/x-canon-crw".into(),
        "dng" => "image/x-adobe-dng".into(),
        "erf" => "image/x-epson-erf".into(),
        "nef" => "image/x-nikon-nef".into(),
        "orf" => "image/x-olympus-orf".into(),
        "rw2" => "image/x-panasonic-rw2".into(),

        // unfortunately I have to return a String, even though everything else is a &'static str
        // because mime_guess returns a lifetime that doesn't last enough, even though they
        // could return a &'static str if they didn't depend on the `mime` crate
        ext => mime_guess::from_ext(ext)
            .first_or_octet_stream()
            .essence_str()
            .into(),
    }
}
```

I think we only need to think if this is the best solution. We already have the mime_type and the extension for formats who also have a matcher, probably the biggest issue here is the `pub(crate) fn no_match(_buf: &[u8]) -> bool { false }` "hack".

Reply:
> I think we only need to think if this is the best solution. We already have the mime_type and the extension for formats who also have a matcher, probably the biggest issue here is the `pub(crate) fn no_match(_buf: &[u8]) -> bool { false }` "hack".

I share your concern about the "hack", but it does the job without the need for further internal refactoring at this stage.
Maybe we can keep it for now until we go further into a mime_guess & infer merge ?

Reply:
I thought about it. What we could do is have the matcher inside infer::Type be an Option that takes None for the extension  only matchers. Then we could put those matchers in a separate file from the magic bytes matchers, so that we could code generate them like mime_guess does without having it be a mess.

Reply:
But still I'm not sure about it, it feels like one crate is doing the work two different crates should do

Reply:
I am totally open to have `mime_guess` address extensions sniffing while `infer` would address content sniffing. In such case we should backout #16 which does not make much sense if we do not merge mime_guess into infer.

Reply:
I would propose then that we move this discussion to a dedicated issue and that I remove the CSS and Javascript support with the `no_match` hack for now, so that we can merge this PR.

In the new issue we will discuss what features should `Infer` provide regarding file extensions and eventually decide to remove `get_from_ext`.

Is that ok for you ?